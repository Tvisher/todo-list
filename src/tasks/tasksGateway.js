const tasksUrlData = 'https://crudcrud.com/api/7998c48afc8540669d6712a0f3c3b628/tasks'

export const createTask = (taskData) => {
    return fetch(tasksUrlData, {
        method: "POST",
        headers: {
            'Content-Type': "application/json,utf-8"
        },
        body: JSON.stringify(taskData)
    }).then(res => {
        if (!res.ok) {
            throw new Error('Failed to create task')
        }
    })
}


export const fetchTaskList = () => {
    return fetch(tasksUrlData)
        .then(res => {
            if (res.ok) {
                return res.json();
            } else {
                throw new Error('Failed to fetch task list')
            }
        }).then(taskList => {
            return taskList.map(({ _id, ...task }) => ({
                id: _id,
                ...task
            }))

        })
}


export const updateTask = (taskId, taskData) => {
    return fetch(`${tasksUrlData}/${taskId}`, {
        method: "PUT",
        headers: {
            'Content-Type': "application/json,utf-8"
        },
        body: JSON.stringify(taskData)
    }).then(res => {
        if (!res.ok) {
            throw new Error('Failed to updated task')
        }
    })
}

export const deleteTask = (taskId) => {
    return fetch(`${tasksUrlData}/${taskId}`, {
        method: "DELETE",
    }).then(res => {
        if (!res.ok) {
            throw new Error('Failed to delete task')
        }
    })
}
