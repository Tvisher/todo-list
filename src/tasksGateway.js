const tasksUrlData = 'https://crudcrud.com/api/e0cdb7b71f36498c942a21b9fdbed5b1/tasks'

export const createTask = (taskData) => {
    return fetch(tasksUrlData, {
        method: "POST",
        headers: {
            'Content-Type': "application/json,utf-8"
        },
        body: JSON.stringify(taskData)
    }).then(res => {
        if (!res.ok) {
            throw new Error('Failed to create task')
        }
    })
}


export const fetchTaskList = () => {
    return fetch(tasksUrlData)
        .then(res => {
            if (res.ok) {
                return res.json();
            } else {
                throw new Error('Failed to fetch task list')
            }
        }).then(taskList => {
            const tasks = taskList.map(({ _id, ...task }) => ({
                id: _id,
                ...task
            }))
            return tasks
        })
}


export const updateTask = (taskId, taskData) => {
    return fetch(`${tasksUrlData}/${taskId}`, {
        method: "PUT",
        headers: {
            'Content-Type': "application/json,utf-8"
        },
        body: JSON.stringify(taskData)
    }).then(res => {
        if (!res.ok) {
            throw new Error('Failed to updated task')
        }
    })
}

export const deleteTask = (taskId) => {
    return fetch(`${tasksUrlData}/${taskId}`, {
        method: "DELETE",
    }).then(res => {
        if (!res.ok) {
            throw new Error('Failed to delete task')
        }
    })
}
