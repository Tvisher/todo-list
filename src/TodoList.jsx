import React, { useState, useEffect } from 'react';
import TasksList from './TasksList'
import CreateTask from './CreateTask';
import { createTask, fetchTaskList, updateTask, deleteTask } from './tasksGateway'

const TodoList = () => {
    const [taskList, setList] = useState([]);

    useEffect(() => {
        fetchTasks()
    }, [])

    const fetchTasks = () => {
        fetchTaskList()
            .then(taskList => setList([...taskList]))
    }

    const addNewTask = (text) => {
        const newTask = {
            text,
            done: false,
        }
        createTask(newTask)
            .then(() => fetchTasks())
    }


    const handleTaskStatusChange = (id) => {
        const { done, text } = taskList.find(task => task.id === id);
        const updatedTusk = {
            text,
            done: !done
        };
        updateTask(id, updatedTusk).then(() => fetchTasks())
    }


    const handleTaskDelete = (id) => {
        deleteTask(id).then(() => fetchTasks())
    }

    return (
        <>
            <h1 className='todo-title'>Todo List</h1>
            <CreateTask addNewTask={addNewTask} />
            <TasksList
                taskList={taskList}
                addNewTask={addNewTask}
                handleTaskStatusChange={handleTaskStatusChange}
                handleTaskDelete={handleTaskDelete} />
        </>
    )
}

export default TodoList;