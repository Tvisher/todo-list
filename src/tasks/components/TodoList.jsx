import React, { useEffect } from 'react';
import TasksList from './TasksList'
import CreateTask from './CreateTask';
import { connect } from 'react-redux'
import * as taskActions from '../tasks.actions'
import { sortedTaskListSelector } from '../tasks.selector'


const TodoList = ({ getTaskList, tasks, updateTaskList, deleteTaskList, createTask }) => {

    useEffect(() => {
        getTaskList()
    }, [getTaskList])


    return (
        <>
            <h1 className='todo-title'>Todo List</h1>
            <CreateTask addNewTask={createTask} />
            <TasksList
                taskList={tasks}
                handleTaskStatusChange={updateTaskList}
                handleTaskDelete={deleteTaskList} />
        </>
    )
}

const mapDispatch = {
    getTaskList: taskActions.getTaskList,
    updateTaskList: taskActions.updateTaskList,
    deleteTaskList: taskActions.deleteTaskList,
    createTask: taskActions.createTask
}

const mapState = state => {
    return {
        tasks: sortedTaskListSelector(state)
    }
}

export default connect(mapState, mapDispatch)(TodoList);