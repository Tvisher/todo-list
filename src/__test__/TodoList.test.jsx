import React from 'react';
import { shallow } from 'enzyme';
import TodoList from '../TodoList';
import { fetchTaskList, deleteTask } from '../tasksGateway'

jest.mock('../tasksGateway.js', () => {
    return {
        createTask: jest.fn(),
        fetchTaskList: jest.fn(() => Promise.resolve([])),
        updateTask: jest.fn(),
        deleteTask: jest.fn(() => Promise.resolve())
    }
})

describe('<TodoList/>', () => {
    it('Should request tasks list', () => {
        shallow(<TodoList />);
        setTimeout(() => {
            expect(fetchTaskList).toBeCalled()
        })
    })

    it('Should handle task delete', () => {
        const wrappedComponent = shallow(<TodoList />);
        const deleteHandler = wrappedComponent.find('TasksList').prop('handleTaskDelete');
        deleteHandler('some-id');
        expect(deleteTask).toBeCalledWith('some-id')
    })
})