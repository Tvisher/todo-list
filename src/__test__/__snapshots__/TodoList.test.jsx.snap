// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TasksList/> Should display sorted by done tasks list 1`] = `
<div
  className="todo-list"
>
  <ul
    className="list"
  >
    <Task
      key="6"
      onChange={[MockFunction]}
      onDelete={[MockFunction]}
      task={
        Object {
          "done": false,
          "id": 6,
          "text": "text 6",
        }
      }
    />
    <Task
      key="5"
      onChange={[MockFunction]}
      onDelete={[MockFunction]}
      task={
        Object {
          "done": false,
          "id": 5,
          "text": "text 5",
        }
      }
    />
    <Task
      key="3"
      onChange={[MockFunction]}
      onDelete={[MockFunction]}
      task={
        Object {
          "done": false,
          "id": 3,
          "text": "text 3",
        }
      }
    />
    <Task
      key="2"
      onChange={[MockFunction]}
      onDelete={[MockFunction]}
      task={
        Object {
          "done": false,
          "id": 2,
          "text": "text 2",
        }
      }
    />
    <Task
      key="4"
      onChange={[MockFunction]}
      onDelete={[MockFunction]}
      task={
        Object {
          "done": true,
          "id": 4,
          "text": "text 4",
        }
      }
    />
    <Task
      key="1"
      onChange={[MockFunction]}
      onDelete={[MockFunction]}
      task={
        Object {
          "done": true,
          "id": 1,
          "text": "text 1",
        }
      }
    />
  </ul>
</div>
`;
